Additional codelists and new study_definition for project

NikNakk opened the pull request:
This is still a work in progress, since it can't be finalised until all of the codelists are in place. I'm therefore leaving this as a draft for now, and this is more a statement of intent rather than an indication that all is ready!

NikNakk committed:
Added OpenSafely IBD code list from 2020-04-07

NikNakk committed:
Added new codelists for Crohn's, UC and IBDU

NikNakk committed:
Merge pull request #1 from opensafely/master

Bring in latest OpenSafely changes

Commit:
Added additional IMID codelists

Commit:
Merge branch 'ibd-codelist'

Commit:
First proper go at study_definition.py

NikNakk committed:
Brought in new codelists

NikNakk committed:
Dummy medication codelists

Also fixed various errors in study definition

NikNakk committed:
Omit updating codelists while waiting for them to be put onto opensafely codelists site

NikNakk committed:
Added AS, HS, ethnicity

NikNakk committed:
Imported HS codes

NikNakk committed:
Split out DMARD code list into separate meds

Also refactored some of study_definition and codelists.py to
avoid repeititive code

NikNakk committed:
Script to avoid update_codelists nuking non-standard codelists

NikNakk committed:
Brought in ado file for table1_mc

Downloaded from http://fmwww.bc.edu/repec/bocode/t/table1_mc.ado

NikNakk committed:
Added new utility functions to make meds easier

The study_definition now has five items for each medication:

- Count from 3m to 0m
- Count from 6m to 3m
- Count from 12m to 6m
- Earliest date since 2010
- Latest date up until 29/02/2020

To avoid repetition, this is implemented by iterating through a list
of tuples and generating a dictionary. The ** unpacking operator is then
used to add this to the study definition.

NikNakk committed:
Removed some duplication of meds

Also added dummy anti-integrin medication list

NikNakk committed:
Changed comments to docstrings

NikNakk committed:
Fixed expected dates for meds

NikNakk committed:
Moved crossimid codelists

Also added placeholders for atopic dermatitis and mesalazine

NikNakk committed:
Removed commenting out of cohort updating

NikNakk commented:
Some codelists are still not finalised. In particular, all of the high cost drug medication lists as well as mesalazine and the atopic dermatitis lists. The atopic dermatitis codes are still also CTV2 since the issue hasn't been updated yet

brianmackenna commented on analysis/codelists.py:
```diff
@@ -0,0 +1,145 @@
+    column="icd10_code",
+)
+
+covid_pos_primcare_code = codelist_from_csv(
```
The current version of protocol says we will only do a sensitivity analysis in SGSS

NikNakk replied:
Ok, we will remove

brianmackenna commented on analysis/codelists.py:
```diff
@@ -0,0 +1,145 @@
+    column="CTV3ID",
+)
+
+covid_pos_primcare_test = codelist_from_csv(
```
As above re sgss

NikNakk replied:
Removed

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+        ("12m_6m", medication_count_12m_6m),
+        ("earliest", medication_earliest),
+        ("latest", medication_latest)
+    ]
```
I think the intention here is to produce 5 different columns here. There are a few things about his approach that we should reconsider
- we don't need all these for every single med
- therefore it doesn't meet principle for data minimisation
- creates extra data which can always add to run times
 - table 4 in protocol specified different dates for each one

NikNakk replied:
I've refined this to two columns for most meds and three for B cell therapies.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+def medication_latest(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2010-01-01", "2020-02-29"],
```
For meds in primary care this will be true but for the secondary care high costs medicines this will only be the last two years

NikNakk replied:
We're no longer pulling the latest date since it's not used in the STATA, but I've left in the function in case it's helpful in future.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+    # Outcomes
+    icu_date_admitted=patients.admitted_to_icu(
+        on_or_after="2020-03-01",
+        include_day=True,
```
we generally don't include the date unless compelling reason for data minimisation principle

NikNakk replied:
We do need the date for the STATA analysis, though it won't be exposed at the end of the STATA script to export

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+    icu_date_admitted=patients.admitted_to_icu(
+        on_or_after="2020-03-01",
+        include_day=True,
+        returning="date_admitted",
```
do we need the date or just binary flag? Not clear from written protocol

NikNakk replied:
We do need the date for the STATA analysis, though it won't be exposed at the end of the STATA script to export

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+        find_first_match_in_period=True,
+        return_expectations={"date": {"earliest": "2020-03-01"}, "incidence": 0.1},
+    ),
+    died_date_cpns=patients.with_death_recorded_in_cpns(
```
Protocol is for only ONS death - delete?

NikNakk replied:
Ok, we will remove

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+        find_first_match_in_period=True,
+        date_format="YYYY-MM-DD",
+        return_expectations={
+            "date": {"earliest": "2020-01-01", "latest": "today"}
```
we discuss using SGSS for a sensitivity analysis but not primary care. Reconsider?

NikNakk replied:
Primary care removed.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
-systolic_blood_pressure_codes = codelist(["2469."], system="ctv3")
-diastolic_blood_pressure_codes = codelist(["246A."], system="ctv3")
+from codelists import *
+
```
We should specify the included population in the python usings `patients.satisfying`. Something like

```
 population=patients.satisfying(
        """
    
        AND (age >=18 AND age <= 110)
        AND has_follow_up 
        AND (sex = "M" OR sex = "F")
        """,
```

NikNakk replied:
Ok, will add.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+        dx_codelist,
+        returning="date",
+        find_first_match_in_period=True,
+        include_month=True,
```
why do we need dates? Is binary yes/no enough? Checking from a data minimisation point of view and I can't see anything in protocol that requires it

NikNakk replied:
We compare dates of diagnosis to determine how to label people where diagnosis has changed.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -1,33 +1,191 @@
+        med_codelist_file = "crossimid-codelists/" + med_codelist_file
+    else:
+        med_codelist_file = "codelists/" + med_codelist_file
+    med_codelist=codelist_from_csv(med_codelist_file + ".csv", system="snomed", column="snomed_id")
```
this is good idea BUT I'm pretty sure we aren't 100% consistent in the meds codelists and there might be further changes with a shift to new BNF coding system.

I should check with team but I think best thing might be for you to run and see hwere it breaks and we fix column names codelist by codelist  on OpenCodelists

NikNakk replied:
Ok. The codelists are consistent for now since most of them were generated from the same source.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -40,70 +198,112 @@
         },
     ),
     # https://github.com/ebmdatalab/tpp-sql-notebook/issues/54
     stp=patients.registered_practice_as_of(
```
stp and msoa re not currently in the protocol.........

NikNakk replied:
We will add STP to the protocol and remove MSOA from the script.

brianmackenna commented on analysis/study_definition.py:
```diff
@@ -121,20 +321,81 @@
         },
     ),
     rural_urban=patients.address_as_of(
```
not in protocol - delete?

NikNakk replied:
IMD is in the protocol. We will remove rural_urban.

brianmackenna commented on codelists/codelists.txt:
```diff
@@ -1,3 +1,27 @@
+opensafely/chronic-respiratory-disease/2020-04-10
+opensafely/current-copd/2020-05-06
+opensafely/covid-identification/2020-06-03
+opensafely/covid-identification-in-primary-care-probable-covid-clinical-code/2020-07-16
+opensafely/covid-identification-in-primary-care-probable-covid-positive-test/2020-07-16
```
comment elsewhere but currently we just have sgss in protocol

NikNakk replied:
We will remove

brianmackenna reviewed (changes requested):
@NikNakk this is a really phenomenal piece of work. Thank you! I've made some comments on the study definition. If you could address these, it will be in good shape for tech team review before merging to master. Just so you are aware your level of Python programming is more advanced than mine so perfectly acceptable to justify and keep.

My comments are generally classed in two areas

-  were we have deviated slightly from the written protocol (it might be that the protocol needs changing)
-  data minimisation - at OpenSAFELY we try to minimise data fields and requests where possible for privacy but also for efficiency reasons.

The [OpenSAFELY documentation](https://docs.opensafely.org/en/latest/) is much more up to date since you wrote this so it might be more helpful now.

I have not reviewed the high cost drugs section.............we should have a better idea on what the code will look like later this week after discussion between OpenSAFELY tech team and James.

NikNakk committed:
Added in high cost drugs

Note this currently doesn't work because of illegal characters in the high cost drug data.

NikNakk committed:
Update crossimid-psoriasis.csv

NikNakk committed:
Removed primary care COVID results

Also:
- Temporary dummy codelist while high cost drug issue being worked on
- Removed earliest and latest med columns since not used downstream
- Removed 6m_12m except for bcell
- Removed medication_counts_and_dates_all since using separate
  function calls permits using named parameters

NikNakk committed:
Merge branch 'master' of github.com:NikNakk/immunosuppressant-meds-research

NikNakk committed:
Removed unneeded data following @brianmackenna's review

NikNakk commented:
I've made the requested changes and/or commented against them. We're now also incorporating high cost drugs, but using a dummy file for now while the issue around invalid characters is sorted out.

NikNakk committed:
Fixed extra population=

NikNakk committed:
Reverted population

(Brian's suggestion not working at present)

NikNakk committed:
Fixed typo

NikNakk committed:
Replaced population definition to include:

- age
- sex
- follow-up

NikNakk committed:
Merge remote-tracking branch 'upstream/master'

brianmackenna commented:
@evansd would you be able to give this code a review please?

I chatted with @sebbacon before Christmas and we think it would be good for someone from tech team to undertake a review before merging.

NikNakk committed:
Remove redundant save_crossimd_codelists script

NikNakk committed:
Removed dummy codelists for high cost drugs

Now that [#453](https://github.com/opensafely-core/cohort-extractor/pull/453) has been fixed,
the dummy codelists are not needed.

evansd commented on save_crossimd_codelists.sh:
```diff
@@ -0,0 +1,6 @@
+mv codelists/crossimid* temp_crossimid/
+cohortextractor update_codelists
+mv temp_crossimid/* codelists/
+rm -r temp_crossimid
```
Is this script still used? It looks like the approach taken above is to keep non-opensafely codelists in a separate `crossimid-codelists` directory, which I think is the correct way of handling this.

The contents of the `codelists` dir needs to match the specification in `codelists.txt` otherwise the tests will fail.

NikNakk replied:
No it's not needed any more. I've removed it with a new commit.

NikNakk replied:
I've also now reverted the code change made to work around the invalid character issue above.

evansd reviewed (approved):
This looks generally good. I don't have enough experience of Stata to sensibly review the Stata code, but the study def code all looks good (and I like the use of functions to remove repetition from the `StudyDefinition` itself).

Now the invalid char issue is fixed (see [#453](https://github.com/opensafely-core/cohort-extractor/pull/453)) it should be possible to include the full high cost drugs codelists, if they're ready.

We've got a new (simpler) test workflow now, so `.github/workflows/test_runner.yaml` should be updated to match [this](https://github.com/opensafely/research-template/blob/master/.github/workflows/test_runner.yaml) and `branch_release_deleter.yaml` is leftover legacy cruft and should be deleted. However, the tests won't run properly on this PR because its from outside of the OpenSAFELY org and so it won't have access to the STATA_LICENSE secret which it needs to run the Stata code. There's a workaround for this which involves someone in the OpenSAFELY org pushing the branch from their account and creating a "shadow PR" which will then cause the test status to update in the original PR.

brianmackenna commented:
I have merged the PR. I have amended write access for @myatesrheum and @NikNakk  to allow direct writing to the repo and run the tests directly.
